core/src/main/clojure/xtdb/expression/temporal.clj

```diff
@@ -930,17 +941,79 @@
                                                   "MILLISECOND" `ChronoUnit/MILLIS
                                                   "MICROSECOND" `ChronoUnit/MICROS))))
                         (zdt->ts ts-unit)))}))
+(defmethod expr/codegen-call [:date_trunc :utf8 :timestamp-tz :utf8] [{[{field :literal} _ {trunc-tz :literal}] :args, [_ [_tstz ts-unit tz :as ts-type] _] :arg-types}]
+  (let [zone-id-sym (gensym 'zone-id)
+        trunc-timezone (ZoneId/of trunc-tz)]
+    {:return-type (conj (pop ts-type) trunc-tz)
+     :batch-bindings [[zone-id-sym `(ZoneId/of ~tz)]]
+     :->call-code (fn [[_ x]] 
+                    (-> `(-> ~(ts->zdt x ts-unit zone-id-sym)
+                             ~(case field
+                                "MILLENNIUM" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (MilleniumAdjuster.)) (.withDayOfYear 1))
+                                "CENTURY" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (CenturyAdjuster.)) (.withDayOfYear 1))
+                                "DECADE" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (DecadeAdjuster.)) (.withDayOfYear 1))
+                                "YEAR" `(-> (.truncatedTo ChronoUnit/DAYS) (.withDayOfYear 1))
+                                "QUARTER" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (QuarterAdjuster.)) (.withDayOfMonth 1))
+                                "MONTH" `(-> (.truncatedTo ChronoUnit/DAYS) (.withDayOfMonth 1))
+                                "WEEK" `(-> (.truncatedTo ChronoUnit/DAYS) (.with (WeekAdjuster.)))
+                                `(.truncatedTo ~(case field
+                                                  "DAY" `ChronoUnit/DAYS
+                                                  "HOUR" `ChronoUnit/HOURS
+                                                  "MINUTE" `ChronoUnit/MINUTES
+                                                  "SECOND" `ChronoUnit/SECONDS
+                                                  "MILLISECOND" `ChronoUnit/MILLIS
+                                                  "MICROSECOND" `ChronoUnit/MICROS)))
+                             (.withZoneSameInstant ~trunc-timezone)) 
```

Comment:
I suspect you'll want to do this TZ change before the truncation, so that it's truncated to (e.g.) the start of the year in the requested TZ?

Reply:
Of course, the example could also be wrong too, it's somewhat hard to really tell!

Reply:
Actually, scratch the above, think you're right